leftwm-core/src/handlers/command_handler.rs

```diff
@@ -275,6 +276,49 @@ fn goto_tag(state: &mut State, input_tag: TagId, current_tag_swap: bool) -> Opti
     state.goto_tag_handler(destination_tag)
 }
 
+fn focus_window(state: &mut State, window_name: &str) -> Option<bool> {
+    let is_target =
+        |w: &Window| -> bool { w.res_name.as_ref().map_or(false, |c| c == window_name) };
+
+    let current_window = state.focus_manager.window(&state.windows)?;
+    let target_window = if is_target(current_window) {
+        let previous_window_handle = state.focus_manager.window_history.get(1);
+        state
+            .windows
+            .iter()
+            .find(|w| Some(&Some(w.handle)) == previous_window_handle)
+            .cloned()
+    } else {
+        state.windows.iter().find(|w| is_target(*w)).cloned()
+    }?;
+
+    let handle = target_window.handle;
+    if target_window.visible() {
+        return Some(handle_focus(state, handle));
+    }
+
+    let tag_id = target_window.tags.first()?;
+    state.goto_tag_handler(*tag_id)?;
+
+    match state
+        .focus_manager
+        .workspace(&state.workspaces)
+        .map(|ws| ws.layout)
+    {
+        Some(Layout::Monocle | Layout::MainAndDeck) => {
+            let mut windows = helpers::vec_extract(&mut state.windows, |w| {
+                w.has_tag(tag_id) && !w.is_unmanaged()
+            });
+            let window_index = windows.iter().position(|w| w.handle == handle)?;
+            let _ = helpers::cycle_vec(&mut windows, -(window_index as i32));
+            state.windows.append(&mut windows);
+            Some(handle_focus(state, handle))
+        }
```

Comment:
I would check whether the window in floating first/whether it is visible again. You will want to exclude floating windows from the extraction as we don't want to cycle through these, the same for the main window in main and deck as this should be fixed.